File: Recusrion/recursion_on_subsequence.py
def rec(idx ,ds:list[int] , a:list[int] , n ):
    
    if idx >= n:
        print(*ds)
        return
    
    ds.append(a[idx])
    rec(idx+1 , ds ,a ,n)
    ds.pop()
    rec(idx+1 , ds ,a ,n)



if __name__ == "__main__":
     
    a = [3,1,2]
    n = 3
    ds = []
    rec(0 , ds , a , n)
     

File: Recusrion/c++/recursion_on_subsequence.cpp
#include<bits/stdc++.h>
using namespace std;
void rec(int idx , vector<int>&ds,int arr[] ,int n)
{  
     if(idx>=n)
     {
        for (auto it : ds)
        {
            cout<<it<<" ";
        }
        cout<<endl;
        return;
     }

     ds.push_back(arr[idx]);
     rec(idx+1,ds,arr,n);
     ds.pop_back();
     rec(idx+1,ds,arr,n);


}
int main(){
     
     int a[]={3,1,2};
     int n = 3;
     vector<int> ds;

     rec(0,ds,a,n);





}

